plugins/ssr/ssrcommon.ml

```diff
@@ -1003,18 +998,29 @@ let dependent_apply_error =
 let applyn ~with_evars ?beta ?(with_shelve=false) ?(first_goes_last=false) n t =
   Proofview.V82.tactic begin fun gl ->
   if with_evars then
-    let refine gl =
-      let t, ty, args, gl = pf_saturate ?beta ~bi_types:true gl t n in
-(*       pp(lazy(str"sigma@saturate=" ++ pr_evar_map None (project gl))); *)
-      let gl = pf_unify_HO gl ty (Tacmach.pf_concl gl) in
-      let gs = CList.map_filter (fun (_, e) ->
-        if EConstr.isEvar (project gl) e then Some e else None)
-        args in
-      pf_partial_solution gl t gs
+    let refine =
+      Proofview.Goal.enter begin fun gl ->
+      Refine.refine ~typecheck:false begin fun sigma ->
+      let env = Proofview.Goal.env gl in
+      let concl = Proofview.Goal.concl gl in
+      let t, ty, args, sigma = saturate ?beta ~bi_types:true env sigma t n in
+      let sigma = unify_HO env sigma ty concl in
+      (* Set our own set of goals. In theory saturate generates them in the
+         right order, so we could just return sigma directly, but explicit is
+         better than implicit. *)
+      let sigma = Evd.push_future_goals (snd @@ Evd.pop_future_goals sigma) in
```

Comment:
Why do we need this belly dance? Why push?

Reply:
As mentioned by the comment we could just return sigma, it works on the ssr test-suite and I believe it's correct. The belly dance is there to remove all the future goals generated by the functions above in the refine block. I could have first pushed at the top then pop here, but since we're in a refine block we now that there is a implicit surrounding push, so popping then pushing is semantially equivalent but more local.

Reply:
Oh, this is morally a "reset"... what an API